pkg/target/target_test.go

```diff
@@ -430,3 +438,144 @@ func TestProcessData(t *testing.T) {
 		})
 	}
 }
+
+var fooMatch = match.Match{
+	Kinds: []match.Kinds{
+		{
+			Kinds:     []string{"Thing"},
+			APIGroups: []string{"some"},
+		},
+	},
+	Scope:              "Namespaced",
+	Namespaces:         []util.Wildcard{"my-ns"},
+	ExcludedNamespaces: nil,
+	LabelSelector: &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"obj": "label",
+		},
+	},
+	NamespaceSelector: &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"ns": "label",
+		},
+	},
+	Name: "",
+}
+
+var fooConstraint = makeConstraint(
+	setKinds([]string{"some"}, []string{"Thing"}),
+	setScope("Namespaced"),
+	setNamespaceName("my-ns"),
+	setLabelSelector("obj", "label"),
+	setNamespaceSelector("ns", "label"),
+)
+
+func TestToMatcher(t *testing.T) {
+	unstructAssign, _ := runtime.DefaultUnstructuredConverter.ToUnstructured(&unversioned.Assign{
+		TypeMeta:   metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{},
+		Spec: unversioned.AssignSpec{
+			Match: fooMatch,
+		},
+	})
+	tests := []struct {
+		name       string
+		constraint *unstructured.Unstructured
+		want       constraints.Matcher
+		wantErr    bool
+	}{
+		{
+			name:       "constraint with no match fields",
+			constraint: makeConstraint(),
+			wantErr:    true,
+		},
+		{
+			name:       "constraint with match fields",
+			constraint: fooConstraint,
+			want: &Matcher{
+				&fooMatch,
+			},
+		},
+		{
+			name: "mutator with match fields",
+			constraint: &unstructured.Unstructured{
+				Object: unstructAssign,
+			},
+			want: &Matcher{
+				&fooMatch,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			h := &K8sValidationTarget{}
+			got, err := h.ToMatcher(tt.constraint)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ToMatcher() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ToMatcher() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestMatcher_Match(t *testing.T) {
+	objData, err := json.Marshal(fooConstraint.Object)
+	if err != nil {
+		t.Fatalf("unable to marshal obj: %s", err)
+	}
+	tests := []struct {
+		name    string
+		match   *match.Match
+		req     interface{}
+		wantErr bool
+	}{
+		{
+			name:    "AdmissionRequest not supported",
+			req:     admissionv1.AdmissionRequest{},
+			wantErr: true,
+		},
+		{
+			name:    "unstructured.Unstructured not supported",
+			req:     makeResource("some", "Thing"),
+			wantErr: true,
+		},
+		{
+			name: "AugmentedReview is supported",
+			req: &AugmentedReview{
+				Namespace: makeNamespace("my-ns"),
+				AdmissionRequest: &admissionv1.AdmissionRequest{
+					Object: runtime.RawExtension{Raw: objData},
+				},
+			},
+			match: &fooMatch,
+		},
+		{
+			name: "AugmentedUnstructured is supported",
+			req: &AugmentedUnstructured{
+				Namespace: makeNamespace("my-ns"),
+				Object:    *fooConstraint,
+			},
+			match: &fooMatch,
+		},
```

Comment:
Add test cases where:

1. the Raw object doesn't unmarshal (e.g. it's invalid JSON)
2. the match logic failes

For 2., it might not yet be possible to induce a failure in the match logic. If that's the case, don't worry about it and we'll add that test case when we add the cache backlink.

Reply:
For 2, It looks like you can induce Match() to fail by making `obj` a Namespace and passing in `nil` for the `ns` argument. This will ensure we test the case where Match() returns an error.

Reply:
"Matching error" case added